### PR TITLE
Add `parallel` and `shuffle` support to `eachobs` and `DataLoader`

### LINKED ISSUE #13: add multitaksing to DataLoader
Port the DataLoader from Flux and extend it with the multitasking features of 
https://github.com/lorenzoh/DataLoaders.jl

### LINKED ISSUE #29: Port collated batch view from DataLoaders.jl
This is the first part of porting functionality from DataLoaders.jl (ref #22).

This includes porting

- `BatchViewCollated` from [`batchview.jl`](https://github.com/lorenzoh/DataLoaders.jl/blob/master/src/batchview.jl)
  - Notes: the `BatchDim` machinery can probably be dropped if we can assume last dim = batch dim
-  `collate` from [`https://github.com/lorenzoh/DataLoaders.jl/blob/master/src/collate.jl`]

`batchview.jl` also includes an unimported `batchsize` helper. Do we want to have a proper `batchsize`?

### LINKED ISSUE #66: API for iterator/view variants
Re #33, what API do we want to go with for the parallel `eachobs`? I think having `eachobs(...; parallel=false)` would be cleanest.

Same question regarding the collated `BatchView` in #63: should we add a `collate` keyword to `BatchView`?

### PR DESCRIPTION
**Pull request by lorenzoh**
Closes #66. Closes #13 

Unifies data iteration under `eachobs` as discussed in https://github.com/JuliaML/MLUtils.jl/issues/66.

- `parallel = true` uses the parallel data loader. `false` by default since parallel does not guarantee correct ordering (see #68)
- `shuffle = true` reshuffles the observations _every time_ you start iterating (unlike `shuffleobs`). This required changing the implementation of non-parallel `eachobs`: they used generator expressions before, but now have a wrapper type, so that we can reach into that when reshuffling. 

With this, I think we can start advertising `eachobs` as the go-to data iteration function. We still need to decide how to proceed with `DataLoader`.

### CODE REVIEW

**Comment by lorenzoh** [comment]
Just saw that `BatchView` does not collate by default, so actually `BatchView` needs collating support before replacing `DataLoader`. But that can be done in another PR.

File: src/eachobs.jl [hunk 1]
```diff
@@ -1,21 +1,29 @@
 """
     eachobs(data, buffer=false, batchsize=-1, partial=true)
```
Comment (CarloLucibello) [inline src/eachobs.jl:2]:
```suggestion
    eachobs(data; [buffer, batchsize, partial, parallel, shuffle])
```
and defaults should be mentioned in the list of arguments

File: src/eachobs.jl [hunk 2]
```diff
@@ -1,21 +1,29 @@
...
-Return an iterator over the observations in `data`. 
+Return an iterator over the observations in `data`.
 
 # Arguments
 
-- `data`. The data to be iterated over. The data type has to implement 
-  [`numobs`](@ref) and [`getobs`](@ref). 
-- `buffer`. If `buffer=true` and supported by the type of `data`, 
+- `data`. The data to be iterated over. The data type has to implement
+  [`numobs`](@ref) and [`getobs`](@ref).
```
Comment (CarloLucibello) [inline src/eachobs.jl:9]:
```suggestion
- `data`. The data to be iterated over. The data type has to be supported by
  [`numobs`](@ref) and [`getobs`](@ref).
```
since `numobs` and `getobs` can support a datatype (through indexing) even if specialized implementations are not given

File: src/eachobs.jl [hunk 3]
```diff
@@ -42,18 +50,80 @@ for (x, y) in eachobs((X, Y))
...
+        else
+            EachObsBuffer(data, buffer)
         end
-        gen = (getobs!(buffer, data, i) for i in 1:numobs(data))
     end
-    return gen
+
+    if shuffle
+        iter = ReshuffleIter(iter)
+    end
```
Comment (CarloLucibello) [inline src/eachobs.jl:79]: shuffling must happen before batching, otherwise mini-batches will always contain the same elements

File: src/eachobs.jl [hunk 4]
```diff
@@ -42,18 +50,80 @@ for (x, y) in eachobs((X, Y))
...
+EachObsBuffer(data) = EachObsBuffer(data, getobs(data, 1))
+Base.length(iter::EachObsBuffer) = numobs(iter.data)
+Base.eltype(iter::EachObsBuffer) = eltype(iter.data)
+
+function Base.iterate(iter::EachObsBuffer)
+    obs = getobs!(iter.buffer, iter.data, 1)
+    return obs, (obs, 2)
+end
+
+function Base.iterate(iter::EachObsBuffer, (buffer, i))
```
Comment (CarloLucibello) [inline src/eachobs.jl:126]:
```suggestion
function Base.iterate(iter::EachObsBuffer, (buffer, i) = (nothing, 1))
```
and remove `iterate(iter)` above

File: src/eachobs.jl [hunk 5]
```diff
@@ -42,18 +50,80 @@ for (x, y) in eachobs((X, Y))
...
+    i > numobs(iter) && return nothing
+    return getobs!(buffer, iter.data, i), (buffer, i+1)
```
Comment (CarloLucibello) [inline src/eachobs.jl:128]: Since `getobs!(nothing, x, i)` is equivalent to `getobs(x, i)`, we could use a unique struct
```julia
struct EachObs
  data
  buffer
end
```
for both buffered and unbuffered iterations

File: src/reshuffle.jl [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    ReshuffleIter(iter)
+
+Wrap a data iterator `iter` created by [`eachobs`](@ref) so that it is
+shuffled anew every time it is iterated over.
+
+This is an internal function. Use `eachobs(data; shuffle = true)` instead.
+"""
+struct ReshuffleIter{T}
+    iter::T
+end
```
Comment (CarloLucibello) [inline src/reshuffle.jl:12]: Do we need this type? Since we are creating a `EachObs` we could store the `shuffle` option in there.

File: src/eachobs.jl (hunk 5 above)
Comment (lorenzoh) [reply src/eachobs.jl:128]: Can do, though I think we would lose the ability to turn off buffering if we just used the same struct for both. So unless I am missing something we would still need to store a flag as to whether buffering is turned on.

If we're doing that, it may make sense to go and add a `parallel` flag as well that is checked when starting to iterate?

File: src/reshuffle.jl (hunk 6 above)
Comment (lorenzoh) [reply src/reshuffle.jl:12]: See above, then I'd suggest to make a single `EachObs` type that holds these flags

File: src/eachobs.jl (hunk 3 above)
Comment (lorenzoh) [reply src/eachobs.jl:79]: True!

File: src/eachobs.jl (hunk 4 above)
Comment (lorenzoh) [reply src/eachobs.jl:126]: Shouldn't this be 
```suggestion
function Base.iterate(iter::EachObsBuffer, (buffer, i) = (iter.buffer, 1))
```
then? So that the stored buffer is used.

**Comment by lorenzoh** [comment]
Based on the feedback, a single `EachObs` would require the least code; what is now being done in the constructor `eachobs` could then be done in the first call to `iterate`, including the reshuflling.

File: src/eachobs.jl (hunk 4 above)
Comment (CarloLucibello) [reply src/eachobs.jl:126]: yes

File: src/eachobs.jl (hunk 5 above)
Comment (CarloLucibello) [reply src/eachobs.jl:128]: > Can do, though I think we would lose the ability to turn off buffering 

Not exactly sure of what you mean here. A call to `eachobs(data, buffer=false)` will produce a `EachObs(data, nothing)` object for which we can `getobs!(iter.buffer, iter.data, i)` falling back to `getobs(iter.data, i)`

File: src/eachobs.jl [hunk 7]
```diff
@@ -42,18 +50,80 @@ for (x, y) in eachobs((X, Y))
...
+            EachObsBuffer(data)
+        else
+            EachObsBuffer(data, buffer)
         end
-        gen = (getobs!(buffer, data, i) for i in 1:numobs(data))
     end
-    return gen
+
+    if shuffle
+        iter = ReshuffleIter(iter)
```
Comment (darsnack) [inline src/eachobs.jl:78]: I would prefer if we allow `EachObs` to contain a field `initialize` (or something) that gets called on the source data before iterating. Similar to how `DataLoader` calls `shuffleobs` but the function can be user specified.

This could allow resampling strategies every pass over the data (e.g. something that comes up in my colleagues research).

**Comment by CarloLucibello** [comment]
> Based on the feedback, a single EachObs would require the least code; what is now being done in the constructor eachobs could then be done in the first call to iterate, including the reshuflling.

And we could rename the `EachObs` in this PR to `DataLoader` and delete the old implementation :exploding_head:

**Comment by CarloLucibello** [comment]
> Is being able to pass a preallocated buffer necessary? 

I'm not sure, I've never used it. But it would be good to avoid breaking changes if not necessary.

File: src/eachobs.jl (hunk 5 above)
Comment (darsnack) [reply src/eachobs.jl:128]: I agree with Carlo about consolidating as much as possible. If we add `buffer` to the regular `EachObs`, then `obs = getobs!(buffer, data, i)` will just use the non-buffered `getobs` when `buffer == nothing` (this is Carlo's point, I think). We shouldn't need a flag cause we just used the stored `iter.buffer` which is `nothing`.

The case of `parallel` seems like it would be harder to consolidate the two structs, so I would prefer to return the appropriate one in `eachobs` like you are now.

File: src/reshuffle.jl (hunk 6 above)
Comment (darsnack) [reply src/reshuffle.jl:12]: I agree, except I would like to expand this a bit. See my comment above.

**Comment by lorenzoh** [comment]
> And we could rename the EachObs in this PR to DataLoader and delete the old implementation 🤯

True, it's looking like `DataLoader` + buffering + parallel now 🙈 . Would you want to keep `eachobs` or `DataLoader` or alias one to the other?

**Comment by lorenzoh** [comment]
> > Is being able to pass a preallocated buffer necessary?
>
> I'm not sure, I've never used it. But it would be good to avoid breaking changes if not necessary.

So which of the three buffers should it accept? Should it validate them? The original docstring also does not say anything about this, so I wonder if it was used:

> - `buffer`. If `buffer=true` and supported by the type of `data`,
a buffer will be allocated and reused for memory efficiency.

File: src/eachobs.jl (hunk 5 above)
Comment (darsnack) [reply src/eachobs.jl:128]: Sorry a bit of a double-post refresh issue!

**Comment by lorenzoh** [comment]
I changed so that it is now just very similar to just `DataLoader` + `parallel` + `buffer`. I can merge one of the two into each other to consolidate. For now, without preallocated buffer, but if we find good semantics for the different argument combinations (see thread above), I can re-add support (at the cost of some bloat).

**Comment by darsnack** [comment]
In terms of the naming, I suggest either consolidating around `DataLoader` or `eachobs` and not having one as convenience for another. If we keep all of them, we don't really solve the confusion from #79 that prompted some of this.

**Comment by darsnack** [comment]
Beyond backward compat, is `batchsize = -1` really necessary? Why not just treat `batchsize = 1` as this case?

**Comment by lorenzoh** [comment]
> Beyond backward compat, is `batchsize = -1` really necessary? Why not just treat `batchsize = 1` as this case?

Batch size 1 is not the same as no batching, e.g. `[obs]` (batch size 1) and `obs` (no batch).

**Comment by lorenzoh** [comment]
> In terms of the naming, I suggest either consolidating around `DataLoader` or `eachobs` and not having one as convenience for another. If we keep all of them, we don't really solve the confusion from #79 that prompted some of this.

I agree. `eachobs` is more idiomatic and in line with the rest of the library while `DataLoader` is easier both for users coming from PyTorch or migrating from DataLoaders.jl.

I was thinking that `DataLoader` could refer to the same function but with better defaults for deep learning, i.e. `parallel = true` and `collate = true` (once added). But then we get into problems with backwards compatibility again.

**Comment by darsnack** [comment]
> Batch size 1 is not the same as no batching, e.g. [obs] (batch size 1) and obs (no batch).

Ah, missed that, thanks.

----

Here's a slightly different implementation that I prefer. I think it offers greater flexibility than boolean switches while still having the user-facing convenience that we want.

```julia
function eachobs(
        data;
        buffer = false,
        parallel = false,
        shuffle = false,
        batchsize::Int = -1,
        partial::Bool = true,
        rng::AbstractRNG = Random.GLOBAL_RNG)
    buffer = buffer isa Bool ? buffer : true

    initialize = if batchsize > 1 && shuffle
        x -> BatchView(shuffleobs(x); batchsize, partial)
    elseif batchsize > 1
        x -> BatchView(x; batchsize, partial)
    elseif shuffle
        shuffleobs
    else
        identity
    end

    return EachObs(data, initialize, buffer, parallel, rng)
end

struct EachObs{T, F, R<:AbstractRNG}
    data::T
    initialize::F
    buffer::Bool
    parallel::Bool
    rng::R
end


function Base.iterate(e::EachObs)
    data = e.initialize(e.data)

    iter = if e.parallel
        eachobsparallel(data; e.buffer)
    else
        if e.buffer
            buf = getobs(data, 1)
            (getobs!(buf, data, i) for i in 1:numobs(data))
        else
            (getobs(data, i) for i in 1:numobs(data))
        end
    end
    obs, state = iterate(iter)
    return obs, (iter, state)
end


function Base.iterate(::EachObs, (iter, state))
    ret = iterate(iter, state)
    isnothing(ret) && return
    obs, state = ret
    return obs, (iter, state)
end


Base.length(e::EachObs) = numobs(e.initialize(e.data))

Base.eltype(e::EachObs) = eltype(e.initialize(e.data))
```

**Comment by darsnack** [comment]
Possibly as a performance optimization, `EachObs` could be mutable and store a cache of `e.initialize(data)`, but at least for the common case variants of `e.initialize`, construction is cheap.

**Comment by lorenzoh** [comment]
@darsnack I think adding an initialization function is a useful addition, but I'm wondering about the semantics. One issue I see with the approach of making shuffling and batching part of this initialization function is that when one wants to add a custom initialization function "on top", you would have to rewrite the shuffling and batching part inside the custom initialization function. Maybe a  `eachobs(data; buffer = false, parallel = false, shuffle = false, batchsize::Int = -1, partial::Bool = true, rng::AbstractRNG = Random.GLOBAL_RNG, `**`initialize = identity`**`)` where `initialize` is applied before shuffling and batching. Additional method with `data(initialize, data; kwargs...)` for `do` syntax could also work then. Could this be adressed in a follow-up PR or would we need breaking changes then?

@CarloLucibello I think it's clear that only one of `eachobs` and `DataLoader` needs to be implemented. Can you give your opinion on which should hold the implementation and what should happen to the other symbol?

**Comment by CarloLucibello** [comment]
As for the naming, `eachobs` is more consistent with the naming convention in the library. 
We cannot remove `DataLoader` since it would break a lot of code, and also having `DataLoader(..)` return an `EachObs` type instead of a `DataLoader` type would be breaking (I myself have some code dispatching on the iterable type). 
So maybe the best solution for the time being is to just go with `DataLoader` and have `eachobs`  as a placeholder:
```julia
eachobs(args...; kws...) = DataLoader(args...; kws...)
```

**Comment by lorenzoh** [comment]
What if we did `const DataLoader = EachObs`. Then I suppose your dispatching code would not break, right?

**Comment by darsnack** [comment]
Re `initialize`: I could see an `initialize = identity` as applying on top of shuffling/batching making sense in the first method that I wrote above. It's easy enough to set `shuffle = false` and `batchsize = -1` to use only the `initialize` part if someone wants to.

As long as we store the shuffling/batching logic within `initialize`, we can always add `eachobs(initialize, data; kwargs...)` later. If `shuffle`, etc. are stored as fields of the struct, then it would be breaking (though `EachObs` itself is un-exported so maybe not?). Personally, I would put everything inside `initialize` since there are few places where we need to replicate the `batchsize > 0` code, so the code is cleaner. But I have no strong feelings on this.

**Comment by CarloLucibello** [comment]
@lorenzoh I took the liberty of pushing this forward and entirely replace the previous DataLoader implementation with this. 
I've been conservative and kept the object name `DataLoader` (I don't want surprises to Flux's users and problems in the docs etc..). We can change the naming scheme in the future.

Right now the only difference between `eachobs` and `DataLoader` is in the default value for `batchsize` (respectively -1 and +1). The -1 default seems more sensible, but I don't want this PR to be breaking.

File: test/eachobs.jl [hunk 8]
```diff
@@ -11,7 +11,7 @@
     for (i,x) in enumerate(eachobs(X, buffer=b))
         @test x == X[:,i]
     end
-    @test b == X[:,end]
+    @test_broken b == X[:,end]
```
Comment (CarloLucibello) [inline test/eachobs.jl:14]: this is the only test that is now failing with this PR. Why is that? It seems we don't use the buffer at all...

File: src/eachobs.jl (hunk 7 above)
Comment (CarloLucibello) [reply src/eachobs.jl]: this can be discussed in a future issue/pr

Comment (darsnack) [reply src/eachobs.jl]: Yeah, that's okay by me.

File: test/eachobs.jl (hunk 8 above)
Comment (CarloLucibello) [reply test/eachobs.jl:14]: @lorenzoh what changed here?

Comment (lorenzoh) [reply test/eachobs.jl:14]: See https://github.com/JuliaML/MLUtils.jl/pull/82#issuecomment-1113995362; I wasn't sure what the right semantics here are, so I didn't implement the preallocated buffer handling. See also https://github.com/JuliaML/MLUtils.jl/pull/82#issuecomment-1114000506.

Comment (lorenzoh) [reply test/eachobs.jl:14]: Is this change okay?

I wouldn't consider this breaking in any case, since `ret = getobs!(buf, data, i)` will fall back to `getobs`, so one should always be using the return value and not rely on `buf` being mutated.

**Review by CarloLucibello — APPROVED** [review]

**Comment by lorenzoh** [comment]
Should I merge?
